producer hanging

Crossposting from Parse.ly's internal issue tracker

### Dec 19:

```
/var/log/cogtree/pixel-producer.log:

 2015-12-19 03:24:02,002 [INFO] blaster.tailer - reopening  /var/log/nginx/plogger
 2015-12-19 03:24:03,649 [INFO] blaster.tailer - Lag behind nginx log is 0:00:01.649717
 2015-12-19 03:24:09,218 [INFO] blaster.tailer - Lag behind nginx log is 0:00:01.218556
 2015-12-19 03:24:14,228 [INFO] blaster.tailer - Lag behind nginx log is 0:00:01.228584
 2015-12-19 03:24:19,691 [INFO] blaster.tailer - Lag behind nginx log is 0:00:03.691249
 2015-12-19 03:24:24,802 [INFO] blaster.tailer - Lag behind nginx log is 0:00:00.802797
 2015-12-19 03:24:30,542 [INFO] blaster.tailer - Lag behind nginx log is 0:00:01.542787
 2015-12-19 03:24:35,564 [INFO] blaster.tailer - Lag behind nginx log is 0:00:00.564231
<hung>
```

### Dec 24:

```
user@server:~$ tail -n50 /var/log/cogtree/pixel-producer.log
2015-12-24 04:06:18,881 [INFO] blaster.tailer - Lag behind nginx log is 0:00:07.881829
2015-12-24 04:06:24,213 [INFO] blaster.tailer - Lag behind nginx log is 0:00:01.213759
2015-12-24 04:06:25,883 [WARNING] pykafka.producer - Broker kafka1:9092 disconnected. Retrying.
2015-12-24 04:06:26,417 [WARNING] pykafka.producer - kafka1:9092 disconnected. Retrying.
2015-12-24 04:06:27,747 [INFO] pykafka.producer - Worker exited for broker kafka1:9092
2015-12-24 04:06:28,059 [ERROR] pykafka.producer - Exception encountered in worker thread:
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/producer.py", line 435, in queue_reader
    self.producer._send_request(batch, self)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/producer.py", line 332, in _send_request
    response = owned_broker.broker.produce_messages(req)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/broker.py", line 257, in produce_messages
    return future.get(ProduceResponse)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/handlers.py", line 59, in get
    raise self.error

2015-12-24 04:06:28,266 [WARNING] pykafka.producer - Broker kafka1:9092 disconnected. Retrying.
Traceback (most recent call last):
  File "/data/apps/cogtree/pixel-producer/publisher.py", line 141, in <module>
    main(settings.TAILER_FILE)
  File "/data/apps/cogtree/pixel-producer/publisher.py", line 136, in main
    producer.produce(str(line))
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/producer.py", line 269, in produce
    self._raise_worker_exceptions()
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/producer.py", line 169, in _raise_worker_exceptions
    raise ex
timeout: timed out
Sentry is attempting to send 1 pending error messages
Waiting up to 10 seconds
Press Ctrl-C to quit
Exception in thread Thread-3 (most likely raised during interpreter shutdown):
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 551, in __bootstrap_inner
  File "/usr/lib/python2.7/threading.py", line 504, in run
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/handlers.py", line 150, in worker
  File "/usr/lib/python2.7/Queue.py", line 174, in get2015-12-24 04:06:30,359 [INFO] blaster.tailer - Connecting to raw_pixels
2015-12-24 04:07:00,389 [ERROR] pykafka.cluster - Unable to connect to broker kafka1:9092
2015-12-24 04:07:00,391 [ERROR] pykafka.cluster - timed out
Traceback (most recent call last):
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/cluster.py", line 229, in _request_metadata
    source_port=self._source_port)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/broker.py", line 92, in __init__
    self.connect()
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/broker.py", line 200, in connect
    self._connection.connect(self._socket_timeout_ms)
  File "/data/virtualenvs/pixel-producer/local/lib/python2.7/site-packages/pykafka/connection.py", line 82, in connect
    (self.source_host, self.source_port)
  File "/usr/lib/python2.7/socket.py", line 571, in create_connection
    raise err
timeout: timed out
2015-12-24 04:07:00,432 [INFO] kazoo.client - Connecting to kafka1:9092
```

These errors appear on machines that produce messages from the us-west to us-east AWS regions and thus have a higher than normal amount of latency. The problem here is that this situation doesn't cause a crash or stacktrace, but only a hang. That makes it impossible for the monitoring daemon to restart these processes.

I've managed to replicate this issue locally by introducing artificial network latency:

    sudo tc qdisc add dev lo root netem delay 8000ms

When I run a producer under these conditions, it quickly gets to a point where it can no longer produce any messages, and even the main thread is hung.

--- a/pykafka/cluster.py
+++ b/pykafka/cluster.py
@@ -259,7 +259,15 @@ class Cluster(object):
             zookeeper = KazooClient(self._seed_hosts,
                                     timeout=self._socket_timeout_ms / 1000)
             try:
-                zookeeper.start()
+                # This math is necessary due to a nested timeout in KazooClient.
+                # KazooClient will attempt to retry its connections only until the
+                # start() timeout is reached. Each of those retries will timeout as
+                # indicated by the KazooClient kwarg. We do a number of timeouts of
+                # self._socket_timeout_ms equal to the number of hosts. This provides
+                # the same retrying behavior that pykafka uses above when treating this
+                # host string as a list of kafka brokers.
+                zookeeper.start(
+                    timeout=(len(broker_connects) * self._socket_timeout_ms) / 1000)
             except Exception as e:
                 log.error('Unable to connect to ZooKeeper instance %s', self._seed_hosts)
                 log.exception(e)

--- a/pykafka/cluster.py
+++ b/pykafka/cluster.py
@@ -222,8 +222,8 @@ class Cluster(object):
         :type broker_connects: Iterable of two-element sequences of the format
             (broker_host, broker_port)
         """
-        try:
-            for host, port in broker_connects:
+        for host, port in broker_connects:
+            try:
                 broker = Broker(-1, host, int(port), self._handler,
                                 self._socket_timeout_ms,
                                 self._offsets_channel_socket_timeout_ms,
@@ -233,9 +233,9 @@ class Cluster(object):
                 response = broker.request_metadata(topics)
                 if response is not None:
                     return response
-        except Exception as e:
-            log.error('Unable to connect to broker %s:%s', host, port)
-            log.exception(e)
+            except Exception as e:
+                log.error('Unable to connect to broker %s:%s. Continuing.', host, port)
+                log.exception(e)
 
     def _get_metadata(self, topics=None):
         """Get fresh cluster metadata from a broker."""
@@ -256,7 +256,8 @@ class Cluster(object):
                 return metadata
 
             # try treating seed_hosts as a zookeeper host list
-            zookeeper = KazooClient(self._seed_hosts, timeout=self._socket_timeout_ms)
+            zookeeper = KazooClient(self._seed_hosts,
+                                    timeout=self._socket_timeout_ms / 1000)
             try:
                 zookeeper.start()
             except Exception as e:

